### PR TITLE
[PS-1755] BEEEP: FSecure FSK-Importer improvements

### PR DESCRIPTION
**Pull request by djsmith85**
## Type of change

```
- [X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Closes https://github.com/bitwarden/clients/issues/2801

I did a fresh export from F-Secure ID protection and used that as a basis for writing new tests and adding types. This was more to ensure there wasn't anything else that might have broke over time.

The importer itself is pretty much unchanged besides fixing #2081 which can be seen [here](https://github.com/bitwarden/clients/commit/eb6faeb4120ec28c8b6f9bde6d4c3e247d3f400d) and [adding support for cipher favorites](https://github.com/bitwarden/clients/commit/55a77e01fa2d030ca44582efb0ed79867feb1630)

## Code changes

- **libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts:** 
  - Moved `fsecureFskImporter` into a new folder `fsecure` under `importers`
  - Renamed it to `fsecure-fsk-importer`
  - Import and use new type `FskFile`
  - Use `type`-property from export to differentiate between cipher types instead of the `style`-property
  - Added support for marking a cipher as favorite on import
  
- **libs/common/spec/importers/testData/fsecureFsk/fsk-test-data.ts:** Added testdata from my export
- **libs/common/spec/importers/fsecure-fsk-importer.spec.ts:** Re-wrote existing tests with new test data
- **libs/common/src/services/import.service.ts:** Fixed import

- **.github/whitelist-capital-letters.txt:** Removed fsecureFskImporter-related files from the linter white-list

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: libs/common/src/importers/fsecureImporters/types/fsecure-fsk-types.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+export interface FskFile {
```
Comment (MGibson1) [inline libs/common/src/importers/fsecureImporters/types/fsecure-fsk-types.ts:1]: Please move up to fsecure-importers folder.

Better to try and not make folders unless it's necessary -- depth ~= complexity. In this case, there are only two files and the separation of the two is clear from the file names. I can't find the blogpost I want to reference, but a good critical mass for creating a (few) subfolder(s) is when there are >~ 3 distinct concerns in the parent and/or 3 files can be joined into a single subfolder concern

File: libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      // eslint-disable-next-line
+      if (!results.data.hasOwnProperty(key)) {
+        continue;
+      }
+
+      const value = results.data[key];
+      const cipher = this.initLoginCipher();
+      cipher.name = this.getValueOrDefault(value.service);
+      cipher.notes = this.getValueOrDefault(value.notes);
+      cipher.favorite = value.favorite > 0;
```
Comment (MGibson1) [inline libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts:28]: Might want to null handle favorite. We're not sure of fsecure's null guarantees.

File: libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+          cipher.card.code = this.getValueOrDefault(value.creditCvv);
+          if (!this.isNullOrWhitespace(value.creditExpiry)) {
+            if (!this.setCardExpiration(cipher, value.creditExpiry)) {
+              this.processKvp(cipher, "Expiration", value.creditExpiry);
+            }
+          }
+          if (!this.isNullOrWhitespace(value.password)) {
+            this.processKvp(cipher, "PIN", value.password);
+          }
+          break;
```
Comment (MGibson1) [inline libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts:51]: Please split type handlers into other methods to appease cyclomatic complexity

File: libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  parse(data: string): Promise<ImportResult> {
+    const result = new ImportResult();
+    const results: FskFile = JSON.parse(data);
+    if (results == null || results.data == null) {
+      result.success = false;
+      return Promise.resolve(result);
+    }
+
+    for (const key in results.data) {
+      // eslint-disable-next-line
```
Comment (MGibson1) [inline libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts:19]: This block can be extracted to a `parseItem` method, again to appease complexity bot

**Review by MGibson1 — CHANGES_REQUESTED** [review]

File: libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
+import { CipherType } from "../../enums/cipherType";
```
[the remark this answers is not in the document]
Comment (djsmith85) [reply libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts:1]: @MGibson1: Should we possibly just drop the `-importers` and just use the vendor-name 

libs/common/src/importers/**fsecureImporters**/fsecure-fsk-importer.ts -> libs/common/src/importers/**fsecure**/fsecure-fsk-importer.ts

I also have some local changes for the enpass importers and had a similar thought.

[the remark this answers is not in the document]
Comment (djsmith85) [reply libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts]: As discussed renamed to use vendor-name

File: libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts (hunk 2 above)
Comment (djsmith85) [reply libs/common/src/importers/fsecureImporters/fsecure-fsk-importer.ts]: Based on the type definition value.favorite is a number and is either 0 or a unix timestamp. Even if favorite is `null` the condition would return false, which would set cipher.favorite to false.

**Review by MGibson1 — DISMISSED** [review]

**Comment by djsmith85** [comment]
@MGibson1 I must have messed up during one of the merges, with the file-rename. Anyhow, both importers and specs were present and guess which one got used when selecting it in the dropdown 🤦‍♂️

**Review by MGibson1 — APPROVED** [review]

**Pull request merged**